### DISCUSSION TITLE
Settings: Ability to move shell commands into different order

### DISCUSSION THREAD

**Discussion by Taitava** [Taitava/obsidian-shellcommands#55] [Ideas]
When I have a lot of shell commands, I'd like to keep certain commands together.

**Comment by FelipeRearden** [comment]
@Taitava , I was thinking about this yesterday but I don't know yet how to do this. 

In my opinion should be a new interface to define this order/groups without messing with the actual interface that works perfect. 
- A new text field inside alias to define a tag ?
- affects only Command Palette order or not?
- A new header on Settings to only view the commands based on these tag?

As a source of information, QuickAdd have a tree view that let's the user aggregate Commands by drag and drop into a category.

![image](https://user-images.githubusercontent.com/85180696/135327385-bd15b77e-e70d-4bfb-9c56-d5e2712f7e23.jpeg)

Let's keep talking :)

**Comment by Taitava** [comment]
> A new text field inside alias to define a tag ?

Did I understand correctly that with *tags*, you mean grouping? This idea would be only about ordering, not defining groups (which could be of course discussed also, but in another idea topic).

> affects only Command Palette order or not?

Unfortunately, I do not know how the command palette is ordered. Commands in command palette seem to appear in a different order than in which order the plugin registers them in Obsidian API. This idea would mainly affect the order in the settings panel, but it might have some consequences in the command panel order, too.

> A new header on Settings to only view the commands based on these tag?

Even when this idea is not about *tags*, some kind of search in the settings could be nice in the future, although not very soon. This can also be discussed in a new idea topic.

> As a source of information, QuickAdd have a tree view that let's the user aggregate Commands by drag and drop into a category.

I also have the idea to use dragging, I just started this topic so quickly between other things that I didn't write much about the idea. There would be a small drag&drop icon at left side of each shell command in the settings.

**Reply by FelipeRearden** [reply]
Sorry @Taitava !!!! I made a lot of mistakes in my previous post :)
- I got too excited and I didn't focus on the central issue: change the order of Shell Commands in Settings tab.

Group Shell Commands is an other issue. You are right :)

> Did I understand correctly that with tags, you mean grouping?

Yes, grouping!

I like this idea too, I am trying to think in a way to implement this in the actual Settings tab

Let keep talking!

**Reply by Taitava** [reply]
Sure! Feel free to post a new topic about the grouping idea sometime. You might have more ideas for it than I do. 👍

**Reply by FelipeRearden** [reply]
Thanks @Taitava !!!! Let me think of a good way to group our beloved Shell Commands :)

**Comment by Taitava** [comment]
[Templater plugin](https://silentvoid13.github.io/Templater/) (version 1.9.8) uses this kind of style to reorder stuff in settings:

![kuva](https://user-images.githubusercontent.com/13002142/135752361-41ebffd9-a88f-4104-bce3-67fdea4912b8.png)

So it has *up* and *down* buttons instead of drag & drop ability. I might consider this too, if drag & drop proves out to be too hard to implement.

**Reply by FelipeRearden** [reply]
Owww much better than drag & drop in my opinion :)

To me, it looks the best approach for Shell Commands :)

**Reply by Taitava** [reply]
I'd think drag & drop would be better for a user, because the user could move a shell command over many commands without the need to repeat a "click *up* button --> move mouse up a bit --> click *up* button again --> move mouse up a bit..." etc. process.

I'm thinking the *up*/*down* button approach as a backup plan.

**Reply by FelipeRearden** [reply]
You are absolutely right @Taitava !!!!
- Specially in our case that have more than 20 shell commands :)

I didn't think about the user operation and I agree with you :)

**Comment by Taitava** [comment]
# More closer for being able to implement this
I'll write just a small heads-up here. #252/#253 will change the structure how shell commands are stored in `data.json` (in version `0.15.0`) in a way that will finally allow reordering them. The user interface for reordering shell commands **is not part of that change**. I just wanted to say that after `0.15.0` is out, it will be easier for me to implement the reordering ability sometime in the future.
